### ISSUE TITLE
Is there some way I can automate this current requirement?

### ISSUE THREAD

**Issue by danielbrunt57** [amitfin/patch#93]
### Checklist

- [X] I have filled out the template to the best of my ability.
- [X] This only contains 1 feature request (if you have multiple feature requests, open one feature request for each feature request).
- [X] This issue is not a duplicate feature request of [previous feature requests](https://github.com/amitfin/patch/issues?q=is%3Aissue+label%3A%22Feature+Request%22+).

### Is your feature request related to a problem? Please describe.

Yes.  Alexa Media Player (`alexa_media`) currently requires `alexapy` 1.29.4.  However, `alexapy` currently requires `aiofiles` 23.x.  HA Core 2024.12.x now uses Python 3.13 with `aiofiles` 24.1.0. 
What I would like to try and automate is the following in the homeassistant container after Core 2024.12 fails to load `alexapy`:
```
pip install alexapy
```
This forces an `aiofiles` downgrade to 23.2.1 but there are two other python libraries in Core 2024.12.x/python 3.13 which require `aiofiles` 24.1.0.  To fix this, reinstall `aiofiles` 24.1.0...
```
pip install aiofiles -U
```
The only component in alexapy which has an issue with Python 3.13 is `alexawebsocket.py`, which I've addressed with a patched file that this repository then replaces for me.

Until I can get alexa_media/alexapy updated, can you suggest any interim ways to accomplish the above mentioned `pip` commands without resorting to command line commands?

Thanks,
Daniel.

### Describe the solution you'd like

See above

### Describe alternatives you've considered

See above

### Additional context

This is really not a feature request!  Just looking for your expert insight!

**Comment by amitfin** [comment]
A possible way to approach it is to prepare a PR with the update for the dependency (starting [here](https://gitlab.com/keatontaylor/alexapy/-/blob/dev/pyproject.toml#L20)). Once the PR is complete (it might require code changes as you mentioned) it's possible to get the list of files to update with their new content.
The `patch` integration allows to use (via patching) the PR before it was accepted.

**Comment by danielbrunt57** [comment]
Thank you for you suggestion!  I'd like to structure it so that it can use aiofiles 23.2.1 or 24.1.0 do that it's not a breaking change but I've not figured out what `aiofiles = "^23.1.0"` should be. And No, I've not googled that yet! The other issue I have though is in alexawebsocket.py where it imports ALLOWED_CLOSE_CODES which fails in 2024.12. I've coded a very crude kludge in my local version of that file but, although it works, it's kind of ugly! Maybe this might have to be a breaking change after all.

So, the patch integration allows to use files from the PR before it was accepted?  Hmmm...I'll have try giving that a go. Up until now, I've only used files from my (config/)patches directory.

**Comment by danielbrunt57** [comment]
Here's my `alexawebsocket.py` "hack":
```
import aiohttp
#from aiohttp import http_websocket
#from aiohttp.http_websocket import WSCloseCode
#from aiohttp.http_websocket.WebSocketReader import ALLOWED_CLOSE_CODES
if sys.version_info.minor == 12:
    from aiohttp.http_websocket import ALLOWED_CLOSE_CODES, WSCloseCode
if sys.version_info.minor == 13:
    from aiohttp.http_websocket import WSCloseCode
    ALLOWED_CLOSE_CODES: Final[Set[int]] = {int(i) for i in WSCloseCode}
    #from aiohttp._websocket.reader_py import ALLOWED_CLODE_CODES
    #from aiohttp.http_websocket.WebSocketReader import ALLOWED_CLOSE_CODES
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding

from alexapy.errors import AlexapyLoginError

from .alexalogin import AlexaLogin  # noqa pylint

_LOGGER = logging.getLogger(__name__)

# Log the result just to make sure
try:
    WSCloseCode
except:
    _LOGGER.error("WSCloseCode is undefined; Python version %s", platform.python_version())
else:
    _LOGGER.debug("WSCloseCode is defined; Python version %s", platform.python_version())

try:
    ALLOWED_CLOSE_CODES
except:
    _LOGGER.error("ALLOWED_CLODE_CODES is undefined; Python version %s", platform.python_version())
else:
    _LOGGER.debug("ALLOWED_CLOSE_CODES is defined; Python version %s", platform.python_version())
```

PS. It's okay to laugh!

**Comment by amitfin** [comment]
This is probably a short term solution and `alexapy` will be updated to support HA 2024.12 in the near future.
If this is true, you can also consider modifying the custom integration `alexa_media` by adding a sub-folder with all of `alexapy`. This way the dependency on `alexapy` 1.29.4 can be avoided and you can modify locally the library as needed.
Once the public `alexapy` library is updated, you can delete `alexa_media` folder and reinstall it (to remove all of the local changes).

**Comment by danielbrunt57** [comment]
> This is probably a short term solution and `alexapy` will be updated to support HA 2024.12 in the near future.

`alexapy` will only be updated if I (or someone else) creates MR and Alan approves it.

> If this is true, you can also consider modifying the custom integration alexa_media by adding a sub-folder with all of alexapy. This way the dependency on alexapy 1.29.4 can be avoided and you can modify locally the library as needed.

I would also have to modify the `alexa_media` `manifest.json` to remove the `alexapy` requirement.
I think I would also have to modify any import references in `alexa_media` and change `alexapy` to `..alexapy` to pull them from the subfolder?

I just finished a session with ChatGPT.  It liked my first suggestion:

> Your suggestion to define ALLOWED_CLOSE_CODES dynamically from WSCloseCode is an elegant and maintainable approach, assuming WSCloseCode is an Enum or iterable containing all valid WebSocket close codes. Here's how you can do it properly:
> 
> Example Definition
> from aiohttp import WSCloseCode
> from typing import Final, Set
> 
> ALLOWED_CLOSE_CODES: Final[Set[int]] = {int(i) for i in WSCloseCode}
 
But I think I'll use this alternative solution for `alexawebsocket.py` which eliminates the need to even try and get ALLOWED_CLOSE_CODES.  It's not quite as "user readable" as before but it also is "future-proof":

```
from typing import Any, Union, cast
#from typing import Any, Final, Set, Union, cast
```
```
import aiohttp
#from aiohttp.http_websocket import ALLOWED_CLOSE_CODES, WSCloseCode
from aiohttp.http_websocket import WSCloseCode
```
```
    def on_close(self, future="") -> None:
        """Handle WebSocket Close."""
        exception_ = self.websocket.exception()
        if exception_:
            self.on_error(str(type(exception_)))
        _LOGGER.debug(
            "WebSocket Connection Closed. %s",
            WSCloseCode(self.websocket.close_code)
            #if self.websocket.close_code in ALLOWED_CLOSE_CODES
            if self.websocket.close_code in WSCloseCode.__members__.values()
            else self.websocket.close_code,
        )
        self._message_count = 0
        asyncio.run_coroutine_threadsafe(self.close_callback(), self._loop)
```

Here is ChatGPT's last comment:

> Is ALLOWED_CLOSE_CODES Necessary?
> While it adds a layer of safety, its utility might seem limited if:

> The WebSocket close codes are expected to always be valid.
> The logs can handle unexpected codes directly (e.g., by logging the raw integer without resolution).
> If the method is used only for debugging, the check could be simplified or removed:

> _LOGGER.debug(
>     "WebSocket Connection Closed. %s",
>     WSCloseCode(self.websocket.close_code)
>     if self.websocket.close_code in WSCloseCode.__members__.values()
>     else self.websocket.close_code,
> )
> Summary
> The check against ALLOWED_CLOSE_CODES ensures:
> 
> Human-readable logging for recognized WebSocket close codes.
> Safety by avoiding exceptions for unrecognized codes.
> If this is the only place it's used and you find the distinction unnecessary, you might simplify the implementation by either removing the ALLOWED_CLOSE_CODES check or redefining its purpose explicitly. Let me know if you need further guidance!

Which approach do you prefer?

**Comment by danielbrunt57** [comment]
Thank you very much for your time!  I really do appreciate it!

This is my `alexapy` merge request: 
[fix: ALLOWED_CLOSE_CODES no longer available in Python 3.13 from aiohttp.http_websocket](https://gitlab.com/keatontaylor/alexapy/-/merge_requests/402#top)

Merry Christmas!
